src/control/server/server_utils.go

```diff
@@ -354,42 +345,40 @@ func prepBdevStorage(srv *server, iommuEnabled bool) error {
 	}
 
 	if bdevCfgs.HaveBdevs() {
-		// The NrHugepages config value is a total for all engines. Distribute allocation
-		// of hugepages across each engine's numa node (as validation ensures that
-		// TargetsCount is equal for each engine). Assumes an equal number of engine's per
-		// numa node.
-		numaNodes, err := getEngineNUMANodes(srv.log, srv.cfg.Engines)
-		if err != nil {
-			return err
-		}
+		if srv.cfg.NrHugepages > 0 {
+			// The NrHugepages config value is a total for all engines. Distribute
+			// allocation of hugepages across each engine's numa node (as validation
+			// ensures that TargetsCount is equal for each engine). Assumes an equal
+			// number of engine's per numa node.
+			numaNodes, err := getEngineNUMANodes(srv.log, srv.cfg.Engines)
+			if err != nil {
+				return err
+			}
 
-		if len(numaNodes) == 0 {
-			return errors.New("invalid number of numa nodes detected (0)")
-		}
+			if len(numaNodes) == 0 {
+				return errors.New("invalid number of numa nodes detected (0)")
+			}
 
-		// Request a few more hugepages than actually required for each NUMA node
-		// allocation as some overhead may result in one or two being unavailable.
-		prepReq.HugepageCount = srv.cfg.NrHugepages / len(numaNodes)
+			// Request a few more hugepages than actually required for each NUMA node
+			// allocation as some overhead may result in one or two being unavailable.
+			prepReq.HugepageCount = srv.cfg.NrHugepages / len(numaNodes)
 
-		// Extra pages to be allocated per engine but take into account the page count
-		// will be issued on each NUMA node.
-		extraPages := (extraHugepages * len(srv.cfg.Engines)) / len(numaNodes)
-		prepReq.HugepageCount += extraPages
-		prepReq.HugeNodes = strings.Join(numaNodes, ",")
+			// Extra pages to be allocated per engine but take into account the page
+			// count will be issued on each NUMA node.
+			extraPages := (extraHugepages * len(srv.cfg.Engines)) / len(numaNodes)
+			prepReq.HugepageCount += extraPages
+			prepReq.HugeNodes = strings.Join(numaNodes, ",")
 
-		srv.log.Debugf("allocating %d hugepages on each of these numa nodes: %v",
-			prepReq.HugepageCount, numaNodes)
-	} else {
-		if srv.cfg.NrHugepages == 0 {
-			// If nr_hugepages is unset then set minimum needed for scanning in prepare
-			// request.
-			prepReq.HugepageCount = scanMinHugepageCount
+			srv.log.Debugf("allocating %d hugepages on each of these numa nodes: %v",
+				prepReq.HugepageCount, numaNodes)
 		} else {
-			// If nr_hugepages has been set manually but no bdevs in config then
-			// allocate on numa node 0 (for example if a bigger number of hugepages are
-			// required in discovery mode for an unusually large number of SSDs).
-			prepReq.HugepageCount = srv.cfg.NrHugepages
+			srv.log.Debugf("skip allocating hugepages, no change is required")
```

Comment:
Should it not be an error to have bdev without huge pages ?

Reply:
After chatting with @NiuYawei decided that we should support emulated NVMe with or without hugepages as some usage models may not require them.

Reply:
Make sense.
NIT, the error `FaultConfigHugepagesDisabledWithBdevs` raised at line should probably be renamed to something such as `FaultConfigHugepagesDisabledWithNvmeBdevs`.

Reply:
fixed